### ISSUE TITLE
Appears to break on windows

### ISSUE THREAD

**Issue by RobertCraigie** [15r10nk/inline-snapshot#43]
Hey! I've recently started migrating my snapshot tests to use `inline-snapshot` - I've been wanting a good inline snapshot library for a while, thank you!

Everything worked almost perfectly for me apart from some [failing tests on windows](https://github.com/RobertCraigie/prisma-client-py/actions/runs/7684886624/job/20941908876?pr=887), here's one of the tracebacks:
```
Traceback (most recent call last):
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\_pytest\pytester.py", line 1172, in runpytest_inprocess
    reprec = self.inline_run(*args, **kwargs)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\_pytest\pytester.py", line 1137, in inline_run
    ret = main([str(x) for x in args], plugins=plugins)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\_pytest\config\__init__.py", line 170, in main
    config=config
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_callers.py", line 116, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\_pytest\main.py", line 318, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\_pytest\main.py", line 307, in wrap_session
    session=session, exitstatus=session.exitstatus
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 418, in traced_hookexec
    return outcome.get_result()
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_result.py", line 108, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_result.py", line 70, in from_call
    result = func()
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 415, in <lambda>
    lambda: oldcall(hook_name, hook_impls, caller_kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_callers.py", line 133, in _multicall
    teardown[0].send(outcome)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\_pytest\terminal.py", line 868, in pytest_sessionfinish
    terminalreporter=self, exitstatus=exitstatus, config=self.config
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 418, in traced_hookexec
    return outcome.get_result()
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_result.py", line 108, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_result.py", line 70, in from_call
    result = func()
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_manager.py", line 415, in <lambda>
    lambda: oldcall(hook_name, hook_impls, caller_kwargs, firstresult)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_callers.py", line 155, in _multicall
    return outcome.get_result()
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_result.py", line 108, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\pluggy\_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\inline_snapshot\pytest_plugin.py", line 153, in pytest_terminal_summary
    unused_externals = _find_external.unused_externals()
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\inline_snapshot\_find_external.py", line 66, in unused_externals
    for name in used_externals():
  File "D:\a\prisma-client-py\prisma-client-py\.nox\test-3-7\lib\site-packages\inline_snapshot\_find_external.py", line 57, in used_externals
    result |= used_externals_in(pathlib.Path(filename).read_text())
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\pathlib.py", line 1222, in read_text
    return f.read()
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 6161: character maps to <undefined>
```

This then cascades into the `pytester` fixture not being able to parse the output.

For reference, this is what one of my failing tests looks like
```py
def test_field_map(testdir: Testdir) -> None:
    def tests() -> None:
        from prisma.models import User

        def test_field_map() -> None:
            user = User(id='1')
            assert user.id == '1'

    testdir.make_from_function(tests)
    testdir.runpytest().assert_outcomes(passed=1)
```

Which is using an internal utility which is just a thin wrapper around pytester's `makefile()` and `runpytest()` methods.

I haven't looked into minimally reproducing this yet.

Note that this is also avoidable by just disabling `inline-snapshot` for the pytester tests, e.g.

```py
pytester.runpytest('-p no:inline_snapshot', *args, **kwargs)
```

**Comment by RobertCraigie** [comment]
Never mind, it appears to break on windows full stop with the same traceback.

**Comment by 15r10nk** [comment]
Thank you for using inline-snapshot and for this report.

I will look into the windows issue and add ci tests if possible.

**Comment by RobertCraigie** [comment]
Thank you! The dynamic snapshots feature is incredible! I haven't seen an inline snapshot implementation that's as flexible before.

**Comment by RobertCraigie** [comment]
Hmm just saw that your PR running windows CI   passes tests, maybe it's the combination of plugins I'm using? Or tests in nested directories? I'll try and narrow it down.

**Comment by 15r10nk** [comment]
I think the problem is that it tries to encode the file with cp1252 (the default encoding on your system). 
I should maybe just use utf-8 to open all files. 
I don't think that it is a problem on your side.

**Comment by RobertCraigie** [comment]
I think your hypothesis is correct! Tests pass after making this change: https://github.com/15r10nk/inline-snapshot/commit/b943d29b3be7f523bdeafc6513b62f987f90e3c9

**Comment by 15r10nk** [comment]
I saw that you use syrupy for snapshots in external files. I'm curious why you don't use [outsource](https://15r10nk.github.io/inline-snapshot/outsource/) to save the snapshots to external files with inline-snapshot?

I only ask because I would like to know a bit more about how people use inline-snapshot.

**Comment by RobertCraigie** [comment]
I was not aware of `outsource()`! I missed it when skimming the readme. I might suggest adding a separate code sample specifically for outsource.

Unfortunately I don't think I'd be able to use it without having more control over the snapshot file name. The main thing I use syrupy for is for snapshots for generated code where the file name it was generated to is very important context.

The snapshots also change very frequently which means the file name would also change frequently which I think wouldn't be a great experience as it'll likely make git diffs & looking through history more challenging.

So an API that allows you to specify the file name, or a pytest option that changes the default behaviour to use the test name would be ideal for me.

```py
assert outsource(codegen, file=Path(__file__).parent.joinpath('__snapshots__/client.py')) == snapshot()
```

**Comment by 15r10nk** [comment]
0.6.1 should fix your issue.

Thank you for your feedback.

The problem with filenames was that they can not be used in every way snapshots can be used.

example:
``` python

global_snapshot=snapshot(external("test_something<the problem>.txt"))

def test_something1():
    ...
    assert outsource("text",file="test_something1.txt") == global_snapshot

def test_something2():
    ...
    assert outsource("text",file="test_something2.txt") == global_snapshot

```

hashes are the only thing which can be used in any situation, where snapshots are used.

But I understand you. I will think about it and provide maybe a way where you can specify the location of the file, or even better generate a location based on the test name and some index which gets incremented.

``` python

def test_something():
    assert outsource("text1", naming="local_file") == snapshot(external("./__snapshots__/test_something_0.txt"))
    assert outsource("text2", naming="local_file") == snapshot(external("./__snapshots__/test_something_1.txt"))
```

But this brings all sorts of other problems:
* you have to rename the snapshots if you rename the test / python file
* snapshots "move" if you remove one, because the index changes
* this can not be combined with every way snapshots can be used (example from above)

I will think about it some time.

I develop this library without much feedback from my users (this is my second issue). I don't know, maybe it is because inline-snapshot is easy to use / maybe I have less users than I think.
I don't know. Thank you for your feedback, it is really valuable for me.

**Comment by RobertCraigie** [comment]
Amazing thanks for the quick fix! Confirmed that everything works with 0.6.1

> you have to rename the snapshots if you rename the test / python file

For what it's worth I think this is a fine tradeoff. I don't think I ever move tests very much in the first place.

> snapshots "move" if you remove one, because the index changes

What if, instead of relying on an index if the same name is used multiple times, you explicitly disallow that in the first place? So if a user wants to use the same name, they'll just have to use an index themselves.

**Comment by 15r10nk** [comment]
Sorry for letting you waiting so long.

I have some plans for inline-snapshot which might not play well with the proposed changes to `outsource`. The key idea of inline-snapshot is that you can see in the source code when your values changed. This works currently also for the outsourced data with hashed names.

My next idea is an browser based snapshot approval-ui #48. Changing how `outsource` works now will make the implementation more difficult.

I will keep your wishes in mind. Maybe I will find a way to make it happen later / maybe there is no reason for it any more.

I will close this issue now because the original bug is fixed.
